packages/react-scheduler/src/ReactScheduler.js

```diff
@@ -300,11 +302,65 @@ if (!ExecutionEnvironment.canUseDOM) {
       isAnimationFrameScheduled = true;
       requestAnimationFrameForReact(animationTick);
     }
-    return newCallbackId;
+    return scheduledCallbackConfig;
   };
 
-  cancelScheduledWork = function(callbackId: number) {
-    delete registeredCallbackIds[callbackId];
+  cancelScheduledWork = function(
+    callbackConfig: CallbackIdType /* CallbackConfigType */,
+  ) {
+    /**
+     * There are four possible cases:
+     * - Head/nodeToRemove/Tail -> null
+     *   In this case we set Head and Tail to null.
+     * - Head -> ... middle nodes... -> Tail/nodeToRemove
+     *   In this case we point the middle.next to null and put middle as the new
+     *   Tail.
+     * - Head/nodeToRemove -> ...middle nodes... -> Tail
+     *   In this case we point the middle.prev at null and move the Head to
+     *   middle.
+     * - Head -> ... ?some nodes ... -> nodeToRemove -> ... ?some nodes ... -> Tail
+     *   In this case we point the Head.next to the Tail and the Tail.prev to
+     *   the Head.
+     */
+    if (callbackConfig.nextCallbackConfig !== null) {
+      // we have a nextCallbackConfig
+      const nextCallbackConfig = callbackConfig.nextCallbackConfig;
+
+      if (callbackConfig.previousCallbackConfig !== null) {
+        // we have a previousCallbackConfig
+        const previousCallbackConfig = callbackConfig.previousCallbackConfig;
+
+        // callbackConfig is somewhere in the middle of a list of 3 or more nodes.
+        previousCallbackConfig.nextCallbackConfig = nextCallbackConfig;
+        nextCallbackConfig.previousCallbackConfig = previousCallbackConfig;
+        return;
+      } else {
+        // there is a nextCallbackConfig but not a previous one;
+        // callbackConfig is the head of a list of 2 or more other nodes.
+        nextCallbackConfig.previousCallbackConfig = null;
+        headOfPendingCallbacksLinkedList = nextCallbackConfig;
+        return;
+      }
+    } else {
+      // there is no next callback config; this must the tail of the list
+
+      if (callbackConfig.previousCallbackConfig !== null) {
+        // we have a previousCallbackConfig
+        const previousCallbackConfig = callbackConfig.previousCallbackConfig;
```

Comment:
And here.

In fact, since we always end up checking these values, we can move the variables to the top of the function.